### PR TITLE
Improve Dehaze for Raw files - Raw black point 

### PR DESCRIPTION
**Pull request by Desmis**
#5769 
Thanks to Ingo Weirich @heckflosse  for this algorithm for Bayer and X-trans sensors which allows in the majority of cases to significantly improve the Dehaze process.

I only added "Dehaze" checkbox in Raw black point (Bayer and X-trans), and of course the whole process to update the sliders. (improccoordinator, rtengine, etc.)

Jacques

### CODE REVIEW

**Comment by Lawrence37** [comment]
Thank you for this @Desmis! It's very easy to use.

One thing I think can be improved is how the adjusters react when dehaze is activated. Right now, changing one of the adjusters causes the image to be reprocessed, then the dehaze reverts the adjusters and the image is reprocessed again. Either the adjusters can be disabled when dehaze is enabled, or changing any of the adjusters should automatically disable dehaze.

**Comment by Desmis** [comment]
@Lawrence37 
I will have a look.

I will make the choice, when Dehaze is enabled to put the 4 (or 3 for RAF) sliders in mode set_sensitive(false).

Thank you for testing

Jacques

**Comment by Lawrence37** [comment]
There is double processing happening when dehaze is enabled for the first time, once after enabling, and once after the adjusters are updated. The preview only shows the dehazed result on the second one, so the double processing is necessary using this implementation. I think it is possible to improve the performance with some small changes.

**Comment by Desmis** [comment]
@Lawrence37 

What do you suggest? You can make the changes without any problem.

Thank you

Jacques

**Comment by Lawrence37** [comment]
I don't know for sure if it will work because I didn't look closely at the code yet, but my idea is to use the calculated values during processing if dehaze is enabled. When the GUI receives the values, it updates the adjusters without sending a ProcEvent.

**Comment by Desmis** [comment]
@Lawrence37 
I remove (with commented  code  and so without suppress code) idle_register in autoblackchanged

Jacques

**Comment by Lawrence37** [comment]
That works. Now I will review the code.

**Comment by marter001** [comment]
As i understand, you put the algorithm of @heckflosse into GUI and this tool works great. Thank you a lot! 
What i observed: Both, in poormandehaze and also in this GUI version the result often comes way too reddish in the darks (for my taste).
Could be, this algorithm isnt meant for landscape pictures at normal sunshine?
grafik

In case you're  interested, i will provide this ARW-file.

**Comment by Desmis** [comment]
@marter001 

Yes, I transformed Ingo's work with the GUI interface.
As Ingo pointed out and as I reported in Pixls.us it doesn't work every time

https://discuss.pixls.us/t/improve-dehaze-for-raw-files/44742

I made the choice - as had been discussed together - to use Raw black point with a checkbox.

It is not recommended otherwise at your own risk to do any retouching, even if it is possible to do so.

Thank you 

Jacques

**Comment by Lawrence37** [comment]
The calculated black values are negative. That doesn't look right.

File: rtdata/languages/default [hunk 1]
```diff
@@ -1404,6 +1404,8 @@ HISTORY_MSG_COLORTONING_LABREGION_SHOWMASK;CT - region show mask
 HISTORY_MSG_COLORTONING_LABREGION_SLOPE;CT - region slope
 HISTORY_MSG_COMPLEX;Wavelet complexity
 HISTORY_MSG_COMPLEXRETI;Retinex complexity
+HISTORY_MSG_DEHABLACK;Dehaze Bayer Black
+HISTORY_MSG_DEHABLACKX;Dehaze X-trans Black
```
Comment (Lawrence37) [inline rtdata/languages/default]:
```suggestion
HISTORY_MSG_DEHABLACKX;Dehaze X-Trans Black
```

**Comment by Lawrence37** [comment]
Somehow the double processing is happening again.

File: rtengine/rawimagesource.cc [hunk 2]
```diff
@@ -2681,8 +2682,22 @@ void RawImageSource::copyOriginalPixels(const RAWParams &raw, RawImage *src, con
...
+*/
+        
+        if (ri->getSensorType() == ST_BAYER) {
+            getMinValsBayer(ri->zeroIsBad());
+        } else {
+            getMinValsXtrans();
+        }
+        std::cout << "red Black : " << minVals[0] << std::endl;
+        std::cout << "green Black : " << minVals[1] << std::endl;
+        std::cout << "blue Black: " << minVals[2] << std::endl;
```
Comment (Lawrence37) [inline rtengine/rawimagesource.cc]: These can be removed now

**Comment by Lawrence37** [comment]
This patch eliminates the double processing and also avoids reprocessing when disabling dehaze, since the black values remain unchanged.

<details>
<summary>diff (click to show/hide)</summary>

```diff
diff --git a/rtengine/improccoordinator.cc b/rtengine/improccoordinator.cc
index 0e6efddce..3e06ca144 100644
--- a/rtengine/improccoordinator.cc
+++ b/rtengine/improccoordinator.cc
@@ -78,7 +78,6 @@ ImProcCoordinator::ImProcCoordinator() :
     sharpMaskChanged(false),
     scale(10),
     highDetailPreprocessComputed(false),
-    computeblack(false),
     highDetailRawComputed(false),
     allocated(false),
     bwAutoR(-9000.f),
@@ -389,7 +388,7 @@ void ImProcCoordinator::updatePreviewImage(int todo, bool panningRelatedChange)
         float greendeha = 0.f;
         float bluedeha = 0.f;
 
-        if ((todo & M_PREPROC)|| (!computeblack) || (!highDetailPreprocessComputed && highDetailNeeded)) {
+        if ((todo & M_PREPROC) || (!highDetailPreprocessComputed && highDetailNeeded)) {
             imgsrc->setCurrentFrame(params->raw.bayersensor.imageNum);
             imgsrc->preprocess(rp, params->lensProf, params->coarse, reddeha, greendeha, bluedeha, true);
             if(imgsrc->getSensorType() == ST_BAYER) {//Bayer
@@ -405,7 +404,6 @@ void ImProcCoordinator::updatePreviewImage(int todo, bool panningRelatedChange)
                     }
                 }
             }
-            computeblack = true;
             if (flatFieldAutoClipListener && rp.ff_AutoClipControl) {
                 flatFieldAutoClipListener->flatFieldAutoClipValueChanged(imgsrc->getFlatFieldAutoClipValue());
             }
diff --git a/rtengine/improccoordinator.h b/rtengine/improccoordinator.h
index de7967984..f8d4424f1 100644
--- a/rtengine/improccoordinator.h
+++ b/rtengine/improccoordinator.h
@@ -92,7 +92,6 @@ protected:
     bool sharpMaskChanged;
     int scale;
     bool highDetailPreprocessComputed;
-    bool computeblack;
     bool highDetailRawComputed;
     bool allocated;
 
diff --git a/rtengine/rawimagesource.cc b/rtengine/rawimagesource.cc
index 12ea1ca7d..83894ac21 100644
--- a/rtengine/rawimagesource.cc
+++ b/rtengine/rawimagesource.cc
@@ -2764,22 +2764,37 @@ void RawImageSource::scaleColors(int winx, int winy, int winw, int winh, const R
 
     if (getSensorType() == ST_BAYER || getSensorType() == ST_FOVEON) {
 
-        black_lev[0] = raw.bayersensor.black1; //R
-        black_lev[1] = raw.bayersensor.black0; //G1
-        black_lev[2] = raw.bayersensor.black2; //B
-        black_lev[3] = raw.bayersensor.black3; //G2
+        if (raw.bayersensor.Dehablack) {
+            black_lev[0] = minVals[0];
+            black_lev[1] = minVals[1];
+            black_lev[2] = minVals[2];
+            black_lev[3] = minVals[1];
+        } else {
+            black_lev[0] = raw.bayersensor.black1; //R
+            black_lev[1] = raw.bayersensor.black0; //G1
+            black_lev[2] = raw.bayersensor.black2; //B
+            black_lev[3] = raw.bayersensor.black3; //G2
+        }
 
         isMono = RAWParams::BayerSensor::getMethodString(RAWParams::BayerSensor::Method::MONO) == raw.bayersensor.method;
     } else if (getSensorType() == ST_FUJI_XTRANS) {
 
-        black_lev[0] = raw.xtranssensor.blackred; //R
-        black_lev[1] = raw.xtranssensor.blackgreen; //G1
-        black_lev[2] = raw.xtranssensor.blackblue; //B
-        black_lev[3] = raw.xtranssensor.blackgreen; //G2  (set, only used with a Bayer filter)
+        if (raw.xtranssensor.Dehablackx) {
+            black_lev[0] = minVals[0];
+            black_lev[1] = minVals[1];
+            black_lev[2] = minVals[2];
+            black_lev[3] = minVals[1];
+        } else {
+            black_lev[0] = raw.xtranssensor.blackred; //R
+            black_lev[1] = raw.xtranssensor.blackgreen; //G1
+            black_lev[2] = raw.xtranssensor.blackblue; //B
+            black_lev[3] = raw.xtranssensor.blackgreen; //G2  (set, only used with a Bayer filter)
+        }
 
         isMono = RAWParams::XTransSensor::getMethodString(RAWParams::XTransSensor::Method::MONO) == raw.xtranssensor.method;
     }
 
+
     for (int i = 0; i < 4 ; i++) {
         cblacksom[i] = max(c_black[i] + black_lev[i], 0.0f);    // adjust black level
     }
diff --git a/rtgui/bayerrawexposure.cc b/rtgui/bayerrawexposure.cc
index d29d8d4b9..0c7a45ebc 100644
--- a/rtgui/bayerrawexposure.cc
+++ b/rtgui/bayerrawexposure.cc
@@ -32,6 +32,7 @@ BayerRAWExposure::BayerRAWExposure () : FoldableToolPanel(this, TOOL_NAME, M("TP
 {
     auto m = ProcEventMapper::getInstance();
     EvDehablack = m->newEvent(DARKFRAME, "HISTORY_MSG_DEHABLACK");
+    EvDehablackVoid = m->newEvent(M_VOID, "HISTORY_MSG_DEHABLACK");
 
     PexBlack1 = Gtk::manage(new Adjuster (M("TP_RAWEXPOS_BLACK_1"), -2048, 2048, 1.0, 0)); //black level
     PexBlack1->setAdjusterListener (this);
@@ -138,37 +139,33 @@ BayerRAWExposure::~BayerRAWExposure()
 void BayerRAWExposure::autoBlackChanged (double reddeha, double greendeha, double bluedeha)
 {
     
- //   idle_register.add(
- //       [this, reddeha, greendeha, bluedeha]() -> bool
+    idle_register.add(
+        [this, reddeha, greendeha, bluedeha]() -> bool
         {
             if (reddeha != PexBlack1->getValue()) {
                 disableListener();
                 PexBlack1->setValue(reddeha);
                 enableListener();
-                adjusterChanged(PexBlack1, 0.);
             }
             if (greendeha != PexBlack0->getValue()) {
                 disableListener();
                 PexBlack0->setValue(greendeha);
                 enableListener();
-                adjusterChanged(PexBlack0, 0.);
             }
             if (greendeha != PexBlack3->getValue()) {
                 disableListener();
                 PexBlack3->setValue(greendeha);
                 enableListener();
-                adjusterChanged(PexBlack3, 0.);
             }
             if (bluedeha != PexBlack2->getValue()) {
                 disableListener();
                 PexBlack2->setValue(bluedeha);
                 enableListener();
-                adjusterChanged(PexBlack2, 0.);
             }
 
-         //   return false;
+            return false;
         }
- //   );
+    );
 }
 
 
@@ -221,7 +218,11 @@ void BayerRAWExposure::checkBoxToggled (CheckBox* c, CheckValue newval)
             PexBlack0->set_sensitive(true);
         }
         if (listener) {
-            listener->panelChanged (EvDehablack, Dehablack->getLastActive() ? M("GENERAL_ENABLED") : M("GENERAL_DISABLED"));
+            if (Dehablack->getLastActive()) {
+                listener->panelChanged (EvDehablack, M("GENERAL_ENABLED"));
+            } else {
+                listener->panelChanged (EvDehablackVoid, M("GENERAL_DISABLED"));
+            }
         }
     }
 }
diff --git a/rtgui/bayerrawexposure.h b/rtgui/bayerrawexposure.h
index d1e03c96c..595b6825d 100644
--- a/rtgui/bayerrawexposure.h
+++ b/rtgui/bayerrawexposure.h
@@ -58,5 +58,6 @@ protected:
     CheckBox* Dehablack;
     IdleRegister idle_register;
     rtengine::ProcEvent EvDehablack;
+    rtengine::ProcEvent EvDehablackVoid;
 
 };
diff --git a/rtgui/xtransrawexposure.cc b/rtgui/xtransrawexposure.cc
index af3786687..de1ae8ca3 100644
--- a/rtgui/xtransrawexposure.cc
+++ b/rtgui/xtransrawexposure.cc
@@ -33,6 +33,7 @@ XTransRAWExposure::XTransRAWExposure () : FoldableToolPanel(this, TOOL_NAME, M("
 {
     auto m = ProcEventMapper::getInstance();
     EvDehablackx = m->newEvent(DARKFRAME, "HISTORY_MSG_DEHABLACKX");
+    EvDehablackxVoid = m->newEvent(M_VOID, "HISTORY_MSG_DEHABLACKX");
     
     PexBlackRed = Gtk::manage(new Adjuster (M("TP_RAWEXPOS_BLACK_RED"), -2048, 2048, 1.0, 0)); //black level
     PexBlackRed->setAdjusterListener (this);
@@ -115,19 +116,16 @@ void XTransRAWExposure::autoBlackxChanged (double reddeha, double greendeha, dou
                 disableListener();
                 PexBlackRed->setValue(reddeha);
                 enableListener();
-                adjusterChanged(PexBlackRed, 0.);
             }
             if (greendeha != PexBlackGreen->getValue()) {
                 disableListener();
                 PexBlackGreen->setValue(greendeha);
                 enableListener();
-                adjusterChanged(PexBlackGreen, 0.);
             }
             if (bluedeha != PexBlackBlue->getValue()) {
                 disableListener();
                 PexBlackBlue->setValue(bluedeha);
                 enableListener();
-                adjusterChanged(PexBlackBlue, 0.);
             }
 
             return false;
@@ -164,7 +162,11 @@ void XTransRAWExposure::checkBoxToggled (CheckBox* c, CheckValue newval)
             PexBlackBlue->set_sensitive(true);
         }
         if (listener) {
-            listener->panelChanged (EvDehablackx, Dehablackx->getLastActive() ? M("GENERAL_ENABLED") : M("GENERAL_DISABLED"));
+            if (Dehablackx->getLastActive()) {
+                listener->panelChanged (EvDehablackx, M("GENERAL_ENABLED"));
+            } else {
+                listener->panelChanged (EvDehablackxVoid, M("GENERAL_DISABLED"));
+            }
         }
     }
 }
diff --git a/rtgui/xtransrawexposure.h b/rtgui/xtransrawexposure.h
index dcd2e8da1..5e82b3158 100644
--- a/rtgui/xtransrawexposure.h
+++ b/rtgui/xtransrawexposure.h
@@ -40,6 +40,7 @@ protected:
     CheckBox* Dehablackx;
     IdleRegister idle_register;
     rtengine::ProcEvent EvDehablackx;
+    rtengine::ProcEvent EvDehablackxVoid;
 
 private:
 //  Gtk::CheckButton*  PextwoGreen;
```

</details>

File: rtdata/languages/default (hunk 1 above)
Comment (Desmis) [reply rtdata/languages/default]: ok

Comment (Desmis) [reply rtdata/languages/default]: ok

File: rtengine/rawimagesource.cc (hunk 2 above)
Comment (Desmis) [reply rtengine/rawimagesource.cc]: ok

**Comment by Desmis** [comment]
@Lawrence37 

I just pushed the change you suggested. Now the system works better and only once. I think it's perfect.

I am always surprised (in a good way) by the finesse of your analysis and your solutions. I looked at the patch and made the changes line by line.

Thank you

Jacques

**Review by Lawrence37 — APPROVED** [review]
Looks good to me!

**Pull request merged**